verl/agent/agent_loop.py

```diff
@@ -0,0 +1,390 @@
+    - Load balance: least requests load balancing
+    - Sticky session: send multi-turn chat completions to same server for automatic prefix caching
+    """
+
+    def __init__(self, config: DictConfig, server_addresses: List[str], max_cache_size: int = 10000):
+        """Initialize the AsyncLLMServerManager.
+
+        Args:
+            config (DictConfig): YAML config.
+            server_addresses (List[str]): OpenAI compatible LLM server addresses.
+            max_cache_size (int, optional): max cache size for request_id to address mapping. Defaults to 10000.
+        """
+        self.config = config
+        self.server_addresses = server_addresses
+
+        # Least requests load balancing
+        self.weighted_addresses = [[0, address] for address in server_addresses]
+        heapq.heapify(self.weighted_addresses)
+
+        # LRU cache to map request_id to address
+        self.request_id_to_address = LRUCache(maxsize=max_cache_size)
+
+    async def chat_completions(
+        self,
+        request_id: str,
+        *,
+        messages: List[Dict[str, Any]],
+        sampling_params: Dict[str, Any],
+        tools: Optional[List[Dict[str, Any]]] = None,
+        **kwargs,
+    ) -> ChatCompletion:
+        """Submit a chat completion request to the LLM server.
+
+        Args:
+            request_id (str): request id for sticky session. In first turn, request_id should be None,
+                and in the following turns, request_id should be ChatCompletion.id.
+            messages (List[Dict[str, Any]]): A list of messages comprising the conversation so far.
+            sampling_params (Dict[str, Any]): Sampling parameters for the chat completion.
+            tools (Optional[List[Dict[str, Any]]], optional): A list of tools to use for the chat completion. Defaults to None.
+            **kwargs: Additional keyword arguments to pass to the chat completion request,
+                refer to https://platform.openai.com/docs/api-reference/chat/create
+
+        Returns:
+            ChatCompletion: The chat completion response from the LLM server.
+        """
+        if request_id:
+            request_id = request_id.removeprefix("chatcmpl-")
+            assert request_id in self.request_id_to_address
+            address = self.request_id_to_address[request_id]
+        else:
+            address = self.weighted_addresses[0][1]
+            self.weighted_addresses[0][0] += 1
+            heapq.heapreplace(self.weighted_addresses, self.weighted_addresses[0])
+
+        # use new request_id to avoid duplicate request_id problem
+        request_id = uuid4().hex
+        self.request_id_to_address[request_id] = address
+
+        return await self._chat_completions(address, request_id, messages=messages, tools=tools, **sampling_params, **kwargs)
+
+    async def _chat_completions(self, address: str, request_id: str, **chat_complete_request) -> ChatCompletion:
```

Comment:
So we finally choose to keep http server routing rather than ray based ?

Reply:
Did you observe a performance diff?

Reply:
It was suffered with concurrency when with sgl-router, and make Zilin Zhu switch to http2 in several weeks before. But may be ok because we only use between each server. 
Another consideration is resource affinity inside message list e.g. tool create resource like image in each turn. 
Using ray will help on auto-affinity scheduling, using http will require user equip with s3 like oss service and cause many network traffic during multi-turn generation. I guess support ray-based routing can improve user experience with less infrastructure requirements.